fluent-dom/src/localization.js

```diff
@@ -0,0 +1,120 @@
+/*eslint no-console: ["error", { allow: ["warn", "error"] }] */
+
+class L10nError extends Error {
+  constructor(message, id, lang) {
+    super();
+    this.name = 'L10nError';
+    this.message = message;
+    this.id = id;
+    this.lang = lang;
+  }
+}
+
+/**
+ * This allows us to cache generated contexts, so that we can generate
+ * them lazily, but then cache them and reuse if we have to retrieve
+ * more entries from that fallback context.
+ */
+
+export class Localization {
+  constructor(id, resIds, generateContexts) {
+    this.id = id;
+    this.resIds = resIds;
+    this.ctxs = generateContexts(resIds);
+  }
+
+  async formatWithFallback(keys, method) {
+    let translations = [];
+    for (let o of this.ctxs) {
+      let ctx = await o.ready();
+      const errors = keysFromContext(method, ctx, keys, translations);
+      if (!errors) {
+        break;
+      }
+    }
+    return translations;
+  }
+
+  formatEntities(keys) {
+    return this.formatWithFallback(keys, messageFromContext);
+  }
+
+
+  formatValues(...keys) {
+    const keyTuples = keys.map(
+      key => Array.isArray(key) ? key : [key, null]
+    );
+    return this.formatWithFallback(keyTuples, valueFromContext);
+  }
+
+  formatValue(id, args) {
```

Comment:
Can you rewrite this to use `async/await`? I'm also considering a rename to `formatString(id, args)`, but I'm fine doing that in a follow-up.